Userland/Games/Sudoku/Board.cpp

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2022, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Board.h"
+#include "AK/Types.h"
+#include "AK/Vector.h"
+#include "LibGfx/Color.h"
+
+Board::Board() { new_game(); }
+
+void Board::new_game()
+{
+
+    // Create fully solved board
+    while (!try_create_board()) { }
+    // Fix extra numbers if needed
+    for (auto i = 0; i < m_number_provided - 17; i++) {
+        auto x = rand() % m_dimension;
+        auto y = rand() % m_dimension;
+        // Find a random square that isn't already fixed
+        while (m_squares[x][y].is_fixed()) {
+            x = rand() % m_dimension;
+            y = rand() % m_dimension;
+        }
+        m_squares[x][y].set_fixed(true);
+    }
+    // Hide the answers but store them to make validation instant
+    for (size_t x = 0; x < m_dimension; x++) {
+        for (size_t y = 0; y < m_dimension; y++) {
+            auto value = m_squares[x][y].get_value();
+            m_squares[x][y].set_answer(value);
+            m_squares[x][y].set_value(0);
+        }
+    }
+}
+
+bool Board::try_create_board()
+{
+    Vector<Vector<Square>> squares;
+
+    squares.resize(m_dimension);
+    for (size_t x = 0; x < m_dimension; x++) {
+        squares[x].resize(m_dimension);
+    }
+
+    // Randomly 17 numbers to the board - the lowest number to provide a unique
+    // solution - https://arxiv.org/pdf/1201.0749.pdf
+    for (auto i = 0; i < 17; i++) {
+        auto x = rand() % m_dimension;
+        auto y = rand() % m_dimension;
+        // Find a random square that doesn't already have a value set
+        while (squares[x][y].get_value()) {
+            x = rand() % m_dimension;
+            y = rand() % m_dimension;
+        }
+
+        auto value = generate_random_value(squares, x, y);
+        if (value == 0) {
+            return false;
+        }
+
+        squares[x][y].set_value(value);
+        squares[x][y].set_fixed(true);
+    }
+    return is_solveable(squares);
+}
+
+bool Board::is_solveable(Vector<Vector<Square>> squares)
```

Comment:
My understanding of pointer may be a little rusty but I don't think it can because it's recursive we need to keep each iteration of the board in memory so we can return to a previous state of the board and try a different path. Whereas if we use pointers here, returning up the tree won't erase the failed attempt.

As the biggest time usage is in memset, I think it would be quicker if didn't do it recursively and instead we were to 'rewind' a single board but I think that's a problem for a future PR

Reply:
Ignore my last comment, I managed to make this change (and fix a terrible bug which was causing the performance issues)